File: trunk/theLink/example/csharp/Filter6.cs
/**
 *  \file       theLink/example/csharp/Filter6.cs
 *  \brief      \$Id$
 *  
 *  (C) 2011 - NHI - #1 - Project - Group
 *  
 *  \version    \$Rev$
 *  \author     EMail: aotto1968 at users.berlios.de
 *  \attention  this software has GPL permissions to copy
 *              please contact AUTHORS for additional information
 */
using System;
using System.IO;
using System.Collections.Generic;
using csmsgque;

namespace example {
  sealed class Filter6 : MqS, IServerSetup, IServerCleanup, IEvent, IService {

    string FN = null;

    public Filter6(MqS tmpl) : base(tmpl) {
    }

    void ErrorWrite () {
      StreamWriter FH = File.AppendText(FN);
      FH.BaseStream.Seek(0, SeekOrigin.End);
      FH.WriteLine("ERROR: " + ErrorGetText());
      FH.Flush();
      FH.Close();
      ErrorReset();
    }

    void LOGF () {
      MqS ftr = ServiceGetFilter();
      if (ftr.LinkGetTargetIdent() == "transFilter") {
	ReadForward(ftr);
      } else {
	//FH = File.AppendText(ReadC());
      FN = ReadC();
      }
      SendRETURN();
    }

    void EXIT () {
      Exit();
    }

    void WRIT () {
      Filter6 master = (Filter6) ServiceGetFilter();
      StreamWriter FH = File.AppendText(master.FN);
      FH.BaseStream.Seek(0, SeekOrigin.End);
      FH.WriteLine(ReadC());
      FH.Flush();
      FH.Close();
      SendRETURN();
    }

    void IService.Service (MqS ctx) {
      StorageInsert();
      SendRETURN();
    }

    void IEvent.Event () {
      if (StorageCount() == 0) {
	ErrorSetCONTINUE();
      } else {
	long Id = 0L;
	try  {
	  MqS ftr = ServiceGetFilter();
	  ftr.LinkConnect();
	  Id = StorageSelect();
	  ReadForward(ftr);
	} catch (Exception ex) {
	  ErrorSet (ex);
	  if (ErrorIsEXIT()) {
	    ErrorReset();
	    return;
	  } else {
	    ErrorWrite();
	  }
	}
	StorageDelete(Id);
      }
    }

    void IServerCleanup.ServerCleanup() {
    }

    void IServerSetup.ServerSetup() {
      Filter6 ftr = (Filter6) ServiceGetFilter();
      ServiceCreate("LOGF", LOGF);
      ServiceCreate("EXIT", EXIT);
      ServiceCreate("+ALL", this);
      ServiceStorage("PRNT");
      ftr.ServiceCreate("WRIT", ftr.WRIT);
      ftr.ServiceProxy("+TRT");
    }

    public static void Main(string[] argv) {
      Filter6 srv = MqFactoryS<Filter6>.Add("transFilter").New();
      try {
	srv.ConfigSetIgnoreExit(true);
	srv.LinkCreate(argv);
	srv.ProcessEvent(MqS.WAIT.FOREVER);
      } catch (Exception ex) {
        srv.ErrorSet (ex);
      }
      srv.Exit();
    }
  }
}

File: trunk/theLink/example/vb/Filter6.vb
REM
REM  \file       theLink/example/vb/Filter6.vb
REM  \brief      \$Id: Filter6.vb 530 2011-03-09 18:03:38Z aotto1968 $
REM  
REM  (C) 2009 - NHI - #1 - Project - Group
REM  
REM  \version    \$Rev: 530 $
REM  \author     EMail: aotto1968 at users.berlios.de
REM  \attention  this software has GPL permissions to copy
REM              please contact AUTHORS for additional information
REM

Imports System
Imports System.IO
Imports csmsgque
Imports System.Collections.Generic

Public Module example
  Private Class Filter6
    Inherits MqS
    Implements IServerSetup
    Implements IEvent
    Implements IService

    Dim FN As String = Nothing

    ' constructor
    Public Sub New(ByVal tmpl As MqS)
      MyBase.New(tmpl)
    End Sub

    Private Sub LOGF()
      Dim ftr As Filter6 = CType(ServiceGetFilter(), Filter6)
      If (ftr.LinkGetTargetIdent() = "transFilter") Then
        ReadForward(ftr)
      Else
        FN = ReadC()
      End If
      SendRETURN()
    End Sub

    Private Sub EXI2()
      [Exit]()
    End Sub

    Private Sub WRIT()
      Dim master As Filter6 = CType(ServiceGetFilter(), Filter6)
      Dim FH As StreamWriter = File.AppendText(master.FN)
      FH.WriteLine(ReadC())
      FH.Flush()
      FH.Close()
      SendRETURN()
    End Sub

    Private Sub ErrorWrite()
      Dim FH As StreamWriter = File.AppendText(FN)
      FH.WriteLine("ERROR: " + ErrorGetText())
      FH.Flush()
      FH.Close()
      ErrorReset()
    End Sub

    Private Sub ServerSetup() Implements IServerSetup.ServerSetup
      Dim ftr As Filter6 = CType(ServiceGetFilter(), Filter6)
      
      ServiceCreate("EXIT", AddressOf EXI2)
      ServiceCreate("LOGF", AddressOf LOGF)
      ServiceCreate("+ALL", Me)
      ServiceStorage("PRNT")
      ftr.ServiceCreate("WRIT", AddressOf ftr.WRIT)
      ftr.ServiceProxy("+TRT")
    End Sub

    Private Sub EventF() Implements IEvent.Event
      If StorageCount() = 0 Then
        ErrorSetCONTINUE()
      Else
        Dim Id As Long = 0L
        Dim ftr As MqS = ServiceGetFilter()
        Try
          ftr.LinkConnect()
          Id = StorageSelect()
          ReadForward(ftr)
        Catch ex As Exception
          ErrorSet(ex)
          If (ErrorIsEXIT()) Then
            ErrorReset()
            Return
          Else
            ErrorWrite()
          End If
        End Try
        StorageDelete(Id)
      End If
    End Sub

    Private Sub Service(ByVal ctx As csmsgque.MqS) Implements IService.Service
      StorageInsert()
      SendRETURN()
    End Sub
  End Class

  Sub Main(ByVal args() As String)
    Dim srv As Filter6 = MqFactoryS(Of Filter6).Add("transFilter").[New]()
    Try
      srv.ConfigSetIgnoreExit(True)
      srv.LinkCreate(args)
      srv.ProcessEvent(MqS.WAIT.FOREVER)
    Catch ex As Exception
      srv.ErrorSet(ex)
    Finally
      srv.Exit()
    End Try
  End Sub
End Module
